bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/types/Type.java

```diff
@@ -31,6 +31,48 @@
  */
 public interface Type {
 
+    // TODO: remove default implementations when standard library types are updated
+    /**
+     * Set the referred type for this type once it has been calculated. This must be called the first time this
+     * calculation is done in order for {@code Type#getReferredTypeCache()} to work properly. This is non-blocking and
+     * will become eventually consistent. Expect {@code TypeUtils#getReferredType(Type)} tobe referentially transparent.
+     *
+     * @param type Type referred by this type. For non-reference types, this is the same type.
+     */
+    default void setCachedReferredType(Type type) {
+    }
+    /**
+     * Get the type referred by this type if it has been already calculated. If it has not been already calculated will
+     * return null. For non-reference types, this will return the same type. This is non-blocking and will become
+     * eventually consistent. Expect {@code TypeUtils#getReferredType(Type)} tobe referentially transparent.
```

Comment:
``` suggestion
     * Get the type referred by this type if it has been already calculated. If it has not been already calculated, it will
     * return null. For non-reference types, this will return the same type. This is non-blocking and will eventually
     * become consistent. Expect {@code TypeUtils#getReferredType(Type)} to be referentially transparent.
```